### PR TITLE
Explorer and toolchain changes to implement #2483

### PR DESCRIPTION
**Pull request by josh11b**
This PR is making two main changes to the Explorer and Toolchain:
- Replace the `is` keyword in `where SomeType is SomeInterface` with `impls`, so it is `where SomeType impls SomeInterface`
- Rewrite uses of the "impls" to something else to avoid, frequently "`impl` declarations" or "implementations", to avoid confusion with the `impls` keyword.

### CODE REVIEW

**Review by jonmeow — APPROVED** [review]
Approving even though the proposal is pending, since leads made a choice and it'll help to just get this in and avoid churn.

However, before you merge, can you please update the PR description to describe the changes being made here, without needing to read other PRs?

**Comment by josh11b** [comment]
> However, before you merge, can you please update the PR description to describe the changes being made here, without needing to read other PRs?

Done.

File: explorer/interpreter/type_checker.cpp
Comment (chandlerc) [inline explorer/interpreter/type_checker.cpp]: This name is a bit confusing because it says `impl` but `constraints`.

But I think this actually points to a whole section of code that might need to be updated, although maybe only minorly? (I also can't drop a comment on those lines, yay.)

Above, there is an `AddImplConstraint` method on the builder. The comments for that seem like they need to be updated (`T is C` and such), but it also isn't completely clear what terminology should be used. On one hand, it is used to track the new `impls` where clause. But it is operating an an `ImplConstraint` type with variables named `impl` etc. My guess is that this section may just need a more general update to pick a consistent set of terminology here and decide whether `ImplConstraint` or something else makes the most sense.

Anyways, this variable name is clearly correct until and unless the rest of the terminology shifts. I'd guess that even the comment updates might make sense as a follow-up anyways to at least fix lingering uses of `is` and if desired adjust terminology or add any explanation of terminology that is useful.

**Review by chandlerc — APPROVED** [review]
Also LG to me fwiw, although I'm less familiar w/ the explorer.

There is a section of code that may need further updates, but I'd not block on that. This PR is big enough and it seems fine to fix forward here. =D

File: explorer/interpreter/type_checker.cpp
Comment (josh11b) [reply explorer/interpreter/type_checker.cpp]: If I understand correctly, you are asking me to replace "impl constraint" with "impls constraint", which I've now done, along with three instances of `T is C` with `T impls C` in comments.

Comment (chandlerc) [reply explorer/interpreter/type_checker.cpp]: The comment changes make sense to me, but the rest probably best for @zygoloid to look at.

File: explorer/interpreter/impl_scope.h
Comment (zygoloid) [inline explorer/interpreter/impl_scope.h]: Hm, these are more general than impl declarations; this also includes things like the "`T` implements `Hashable`" fact we get from `fn F[T:! Hashable](x: T)` or from `fn F[T:! type where T impls Hashable](x: T)`.

I don't have a great name for these in mind, but I think `impl_declarations` is probably too confusing. The least-worst idea I have so far is to rename class `Impl` to `ImplFact` and call this `impl_facts`. But I don't like it!

File: explorer/interpreter/type_checker.cpp
Comment (zygoloid) [inline explorer/interpreter/type_checker.cpp:1704]: Maybe `num_equals_constraints_added`, given that we call the representation an `EqualsConstraint`.

File: explorer/interpreter/impl_scope.h
Comment (josh11b) [reply explorer/interpreter/impl_scope.h]: I've made that change.

File: explorer/interpreter/impl_scope.h
Comment (zygoloid) [inline explorer/interpreter/impl_scope.h:57]: Comment on line 46 needs updating.

**Review by zygoloid — APPROVED** [review]
Looks good to me.

**Pull request merged**
